rc/tools/lint.kak

```diff
@@ -1,113 +1,339 @@
+                gsub(/\x27/, "\x27\x27", text)
+                return "\x27" text "\x27"
+            }
 
-        # Flags for the gutter:
-        #   stamp l3|{red}█ l11|{yellow}█
-        # Contextual error messages:
-        #   stamp 'l1.c1,l1.c1|kind:message' 'l2.c2,l2.c2|kind:message'
-        awk -F: -v file="$kak_buffile" -v stamp="$kak_timestamp" -v client="$kak_client" '
             BEGIN {
+                OFS=":"
+                FS=":"
                 error_count = 0
                 warning_count = 0
             }
-            /:[1-9][0-9]*:[1-9][0-9]*: ([Ff]atal )?[Ee]rror/ {
-                flags = flags " " $2 "|{red}█"
-                error_count++
-            }
+
             /:[1-9][0-9]*:[1-9][0-9]*:/ {
-                if ($4 !~ /[Ee]rror/) {
-                    flags = flags " " $2 "|{yellow}█"
+                # Remember that an error or a warning occurs on this line..
+                if ($4 ~ /[Ee]rror/) {
+                    # We definitely have an error on this line.
+                    flags_by_line[$2] = "{Error}!"
+                    error_count++
+                } else if (flags_by_line[$2] ~ /Error/) {
+                    # We have a warning on this line,
+                    # but we already have an error, so do nothing.
+                    warning_count++
+                } else {
+                    # We have a warning on this line,
+                    # and no previous error.
+                    flags_by_line[$2] = "{Information}?"
                     warning_count++
                 }
-            }
-            /:[1-9][0-9]*:[1-9][0-9]*:/ {
-                kind = substr($4, 2)
-                error = $2 "." $3 "," $2 "." $3 "|" kind
-                msg = ""
-                # fix case where $5 is not the last field because of extra colons in the message
+
+                # The message starts with the severity indicator.
+                msg = substr($4, 2)
+
+                # fix case where $5 is not the last field
+                # because of extra colons in the message
                 for (i=5; i<=NF; i++) msg = msg ":" $i
+
+                # Mention the column where this problem occurs,
+                # so that information is not lost.
+                msg = msg "(col " $3 ")"
+
+                # FIXME: I *think* this line is left over from a time
+                # when Kakoune used pipes to delimit items in a list
+                # option. If so, it is entirely useless now... but I
+                # am not 100% sure.
                 gsub(/\|/, "\\|", msg)
-                gsub("'\''", "'"''"'", msg)
-                error = error msg " (col " $3 ")"
-                errors = errors " '\''" error "'\''"
+
+                if ($2 in messages_by_line) {
+                    # We already have a message on this line,
+                    # so append our new message.
+                    messages_by_line[$2] = messages_by_line[$2] "\n" msg
```

Comment:
Currently there's no way for a linter to produce a multi-line message, since any extra context or information is dropped by `awk`). If we ever support multi-line messages, it would be nice to put a bullet at the beginning of each message (and whitespace at the beginning of line which are not the first line of a message).

That's a future improvement, though.

Reply:
I meant that showing a list of one-line diagnostics can be confusing, if there's no way to know they might be unrelated with each other - the user will read the whole thing as one paragraph. Handling dupe might be a good future improvement as well.

Reply:
Right now, every line starts with "warning:" or "error:" and ends with "(col 27)" (or whatever column, so I think the risk of reading the entire info block as a single paragraph is slim. Are you OK with leaving that formatting to a future PR?

Reply:
If there's a prefix to every line then never mind!